File: app/src/main/java/tech/pauly/findapet/utils/ExpandingLayout.java
package tech.pauly.findapet.utils;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.animation.FastOutSlowInInterpolator;
import android.util.AttributeSet;
import android.view.ViewTreeObserver;
import android.widget.LinearLayout;

import tech.pauly.findapet.R;

public class ExpandingLayout extends LinearLayout {

    private static final int DEFAULT_DURATION = 500;

    private enum State {
        COLLAPSED,
        COLLAPSING,
        EXPANDED,
        EXPANDING
    }

    private final FastOutSlowInInterpolator interpolator = new FastOutSlowInInterpolator();
    private ValueAnimator animator;

    private int duration;
    private State currentState = State.COLLAPSED;
    private float currentAnimationValue = 0;
    private int fullLayoutHeight = 0;
    private boolean forceCollapse = false;

    public ExpandingLayout(@NonNull Context context) {
        super(context);
        setup(context, null);
    }

    public ExpandingLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setup(context, attrs);
    }

    public ExpandingLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setup(context, attrs);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int expansionSize = (int) (fullLayoutHeight * currentAnimationValue);
        setMeasuredDimension(getMeasuredWidth(), expansionSize);
    }

    public void collapse() {
        if (currentState != State.COLLAPSED) {
            forceCollapse = true;
            tapToggle();
        }
    }

    public void tapToggle() {
        if (animator != null) {
            animator.cancel();
            animator = null;
        }

        if (forceCollapse || currentState == State.EXPANDING || currentState == State.EXPANDED) {
                animator = ValueAnimator.ofFloat(currentAnimationValue, 0);
        } else {
            animator = ValueAnimator.ofFloat(currentAnimationValue, 1);
        }
        forceCollapse = false;

        animator.setDuration(duration)
                .setInterpolator(interpolator);
        animator.addUpdateListener(valueAnimator -> animate((float) valueAnimator.getAnimatedValue()));
        animator.start();
    }

    private void animate(float animationValue) {
        float lastAnimationValue = currentAnimationValue;
        currentAnimationValue = animationValue;

        if (currentAnimationValue == 0) {
            currentState = State.COLLAPSED;
        } else if (currentAnimationValue == 1) {
            currentState = State.EXPANDED;
        } else {
            if (currentAnimationValue > lastAnimationValue) {
                currentState = State.EXPANDING;
            } else {
                currentState = State.COLLAPSING;
            }
        }

        setVisibility(currentState == State.COLLAPSED ? GONE : VISIBLE);
        requestLayout();
    }

    private void setup(@NonNull Context context, @Nullable AttributeSet attrs) {
        if (attrs != null) {
            TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ExpandingLayout);
            try {
                duration = a.getInt(R.styleable.ExpandingLayout_duration, DEFAULT_DURATION);
            } finally {
                a.recycle();
            }
        }
        measureLayoutHeight();
    }

    private void measureLayoutHeight() {
        getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                for (int i = 0; i < getChildCount(); i++) {
                    fullLayoutHeight += getChildAt(i).getHeight();
                }
                getViewTreeObserver().removeOnGlobalLayoutListener(this);
                setVisibility(GONE);
            }
        });
    }
}

File: app/src/main/java/tech/pauly/findapet/utils/ExpandingLayout.kt
package tech.pauly.findapet.utils

import android.animation.ValueAnimator
import android.content.Context
import android.support.v4.view.animation.FastOutSlowInInterpolator
import android.util.AttributeSet
import android.view.View
import android.view.ViewTreeObserver
import android.widget.LinearLayout
import tech.pauly.findapet.R

private const val DEFAULT_DURATION = 500
private const val COLLAPSED_VALUE = 0f
private const val EXPANDED_VALUE = 1f

class ExpandingLayout : LinearLayout {

    private enum class State {
        COLLAPSED,
        COLLAPSING,
        EXPANDED,
        EXPANDING
    }

    private val interpolator = FastOutSlowInInterpolator()
    private var animator: ValueAnimator? = null

    private var duration = 0L
    private var currentState = State.COLLAPSED
    private var currentAnimationValue = COLLAPSED_VALUE
    private var fullLayoutHeight = 0
    private var forceCollapse = false

    constructor(context: Context) : super(context) {
        setup(context, null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        setup(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        setup(context, attrs)
    }

    private fun setup(context: Context, attrs: AttributeSet?) {
        attrs?.apply {
            val a = context.obtainStyledAttributes(this, R.styleable.ExpandingLayout)
            try {
                duration = a.getInt(R.styleable.ExpandingLayout_duration, DEFAULT_DURATION).toLong()
            } finally {
                a.recycle()
            }
        }
        measureLayoutHeight()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val expansionSize = (fullLayoutHeight * currentAnimationValue).toInt()
        setMeasuredDimension(measuredWidth, expansionSize)
    }

    fun collapse() {
        if (currentState != State.COLLAPSED) {
            forceCollapse = true
            tapToggle()
        }
    }

    fun tapToggle() {
        animator?.let {
            it.cancel()
            animator = null
        }

        animator = if (forceCollapse || currentState == State.EXPANDING || currentState == State.EXPANDED) {
            ValueAnimator.ofFloat(currentAnimationValue, COLLAPSED_VALUE)
        } else {
            ValueAnimator.ofFloat(currentAnimationValue, EXPANDED_VALUE)
        }.also {
            forceCollapse = false
            it.setDuration(duration).interpolator = this.interpolator
            it.addUpdateListener { valueAnimator -> animate(valueAnimator.animatedValue as Float) }
            it.start()
        }
    }

    private fun animate(animationValue: Float) {
        val lastAnimationValue = currentAnimationValue
        currentAnimationValue = animationValue

        currentState = when {
            currentAnimationValue == COLLAPSED_VALUE -> State.COLLAPSED
            currentAnimationValue == EXPANDED_VALUE -> State.EXPANDED
            currentAnimationValue > lastAnimationValue -> State.EXPANDED
            else -> State.COLLAPSING
        }

        visibility = if (currentState == State.COLLAPSED) View.GONE else View.VISIBLE
        requestLayout()
    }

    private fun measureLayoutHeight() {
        viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                for (i in 0 until childCount) {
                    fullLayoutHeight += getChildAt(i).height
                }
                viewTreeObserver.removeOnGlobalLayoutListener(this)
                visibility = View.GONE
            }
        })
    }
}
